clang/lib/CIR/CodeGen/LoweringPrepare.cpp

```diff
@@ -865,30 +893,379 @@ void LoweringPreparePass::lowerThreeWayCmpOp(CmpThreeWayOp op) {
+  assert(!cir::MissingFeatures::functionIndexAttribute());
+  assert(!cir::MissingFeatures::noUnwindAttribute());
+  return cgm.createRuntimeFunction(fTy, "__cxa_guard_acquire");
+}
+
+static cir::FuncOp getGuardReleaseFn(clang::CIRGen::CIRGenModule &cgm,
+                                     cir::PointerType guardPtrTy) {
+  // void __cxa_guard_release(__guard *guard_object);
+  cir::FuncType fTy = cgm.getBuilder().getFuncType(guardPtrTy, {cgm.VoidTy},
+                                                   /*isVarArg=*/false);
+  assert(!cir::MissingFeatures::functionIndexAttribute());
+  assert(!cir::MissingFeatures::noUnwindAttribute());
+  return cgm.createRuntimeFunction(fTy, "__cxa_guard_release");
+}
+
+namespace {
+struct CallGuardAbort final : clang::CIRGen::EHScopeStack::Cleanup {
+  cir::GlobalOp guard;
+  CallGuardAbort(cir::GlobalOp guard) : guard{guard} {}
+
+  void Emit(clang::CIRGen::CIRGenFunction &cgf, Flags flags) override {
+    auto guardPtrTy = cir::PointerType::get(guard.getSymType());
+    mlir::Value getGuard =
+        cgf.CGM.getBuilder().createGetGlobal(guard, /*threadLocal*/ false);
+    cgf.emitNounwindRuntimeCall(guard->getLoc(),
+                                getGuardAbortFn(cgf.CGM, guardPtrTy), getGuard);
+  }
+};
+} // namespace
+
+void LoweringPreparePass::handleStaticLocal(GlobalOp globalOp,
+                                            GetGlobalOp getGlobalOp) {
+
+  std::optional<cir::ASTVarDeclInterface> astOption = globalOp.getAst();
+  assert(astOption.has_value());
+  cir::ASTVarDeclInterface varDecl = astOption.value();
+
+  clang::CIRGen::CIRGenModule &cgm = *this->cgm;
+  clang::CIRGen::CIRGenCXXABI &cxxABI = cgm.getCXXABI();
+  clang::CIRGen::CIRGenBuilderTy &builder = cgm.getBuilder();
+
+  builder.setInsertionPointAfter(getGlobalOp);
+  Block *getGlobalOpBlock = builder.getInsertionBlock();
+  // TODO(CIR): This is too simple at the moment. This is only tested on a
+  // simple test case with only the static local var decl and thus we only have
+  // the return. For less trivial examples we'll have to handle shuffling the
+  // contents of this block more carefully.
+  Operation *ret = getGlobalOpBlock->getTerminator();
+  ret->remove();
+  builder.setInsertionPointAfter(getGlobalOp);
+
+  clang::CIRGen::CIRGenFunction cgf{cgm, builder, true};
+  cgf.CurFn = getGlobalOp->getParentOfType<cir::FuncOp>();
+  cgm.setCurrentCIRGenFn(&cgf);
+
+  // Inline variables that weren't instantiated from variable templates have
+  // partially-ordered initialization within their translation unit.
+  bool nonTemplateInline =
+      varDecl.isInline() &&
+      !isTemplateInstantiation(varDecl.getTemplateSpecializationKind());
+
+  // We only need to use thread-safe statics for local non-TLS variables and
+  // inline variables; other global initialization is always single-threaded
+  // or (through lazy dynamic loading in multiple threads) unsequenced.
+  bool threadsafe = astCtx->getLangOpts().ThreadsafeStatics &&
```

Comment:
If we need all these AST queries to compute threadSafe, looks like this could be unitAttr computed during CIRGen? Like you mentioned, seems like the added interfaces are a little bit too much